### PR TITLE
Search functionality

### PR DESCRIPTION
**Pull request by glasnt**
Adds a search box to the navbar with the dates. Uses the same threading method as dates, including header-link data transfer, to limit the results of any page. 

 * matches on incident_key, non case sensitive
 * wedged in as a precondition into existing searches
 * navbar form altered so clicking 'load' or hitting enter on field uses the parameters for the new search.

### CODE REVIEW

File: pigeonhole.rb [hunk 1]
```diff
@@ -30,6 +30,14 @@
   redirect "/noise-candidates/#{today}/#{today}"
 end
 
+def search_query_input 
+  if @search
```
Comment (Hobbsee) [inline pigeonhole.rb]: Cleaner as 
return {} unless @search
{ :conditions => "and incident_key =~ /.*#{@search.strip}.*/i" }

File: views/noise-candidates.haml [hunk 2]
```diff
@@ -8,6 +8,9 @@
     = @start_date
     and
     = @end_date
+    - if @search
```
Comment (Hobbsee) [inline views/noise-candidates.haml]: This is indented too far, and won't show in the case where the start date is the same as the end date.

**Comment by Hobbsee** [comment]
I would prefer to see search_query_input used everywhere, rather than a mixture of search_query_input, passed_query_input, query_input, and precondition.  

In fact, I'm tempted to say that all sanitising should be done in the def search_query_input, and have that function return an empty string or a the value you currently have in :conditions in the hash, so it can be used directly.

**Comment by Hobbsee** [comment]
As an aside, what sanitising do we need to do here?

File: pigeonhole.rb (hunk 1 above)
Comment (nerdyness) [inline pigeonhole.rb]: Really? Is this opinionated or what makes this cleaner?
The style-guide says to prefer this:
`@search ?  { :conditions => "and incident_key =~ /.*#{@search.strip}.*/i" } : {}`

Comment (Hobbsee) [inline pigeonhole.rb]: That's better again, I just didn't think of it.  Thanks!

**Comment by nerdyness** [comment]
> As an aside, what sanitising do we need to do here?

Any destructive instructions I guess. `DELETE from timeseries;` and the like.

We could seriously minimise damage if we allowed for a read-only user and a more privileged read/write user here. The import script uses the read/write user and hitting submit on the categorisation page opens a new influx connection using the read/write user credentials. Everything else, especially the search runs as the read-only user.

So the worst we'd have to fear is that somebody runs a `select * from timeseries;` and gets all information they could have queried anyway :smile:

**Comment by glasnt** [comment]
@Hobbsee your inline comments have been addressed. 

Re: search_query_input: this has been converted to a single string that is either prepended to a conditional hash value, or used as the value for a new hash being passed in (because we call find_incidents from the main file, as well as inside other calls)

Re: SQL sanitization (@nerdyness) : we could start stripping common injection vectors (semicolons, hyphens, etc); but I think limiting the permissions of the influx user in the app is a better mitigation technique.

**Comment by glasnt** [comment]
On latest commit: a brief statistical analysis shows that of the data being used in testing, there are cases where the generated incident key uses colons and hyphens. The latest change removes quotes (both kinds) and semicolons from the global variable itself, then passes it down. This means that the elements in the field itself will be removed. They will still appear, URI-encoded, in the URL. 

I had hoped that the influx query function, and the fact that the admin interface does say 'Read points' for queries meant that it would only accept 'select' commands anyway. However, the fact that I was able to delete all my local data just by running `delete from pigeonhole` in this 'read' query interface means that no, it does allow data manipulation commands.

**Comment by nerdyness** [comment]
> The latest change removes quotes (both kinds) and semicolons from the global variable itself

That is IMO the wrong approach. Those characters should be escaped, not removed.

**Comment by glasnt** [comment]
I can address this. The main problem is encoding for influx and
sinatra/rack and the uri

**Comment by glasnt** [comment]
Functionality changed to escaping, @nerdyness

**Comment by nerdyness** [comment]
> Functionality changed to escaping, @nerdyness

:+1: Have you tested that `URI.escape` works as expected using influxdb? I hear it's not your average database and you shouldn't trust it :wink:

**Comment by glasnt** [comment]
I think it'll work well. I haven't done a full penetration test on that vector, but from what I can see, it should be ok.

**Comment by nerdyness** [comment]
> I think it'll work well. I haven't done a full penetration test on that vector, but from what I can see, it should be ok.

Good to hear, thanks @glasnt!

File: pigeonhole.rb [hunk 3]
```diff
@@ -41,14 +48,17 @@
   ]
   @start_date = params["start_date"]
   @end_date   = params["end_date"]
-  @incidents = influxdb.find_incidents(@start_date, @end_date)
+  @search     = params["search"]
+puts search_precondition
```
Comment (Hobbsee) [inline pigeonhole.rb]: Debug code left in.

Comment (glasnt) [inline pigeonhole.rb]: Nice catch. Fixed.

File: lib/influx.rb [hunk 4]
```diff
@@ -98,8 +98,8 @@ def incident_frequency(start_date = nil, end_date = nil)
       }.compact
     end
 
-    def alert_response(start_date = nil, end_date = nil)
-      incidents = find_incidents(start_date, end_date)
+    def alert_response(start_date = nil, end_date = nil, precondition = "")
+      incidents = find_incidents(start_date, end_date, { :conditions => precondition })
```
Comment (Hobbsee) [inline lib/influx.rb]: The {}s are redundant here.

**Comment by Hobbsee** [comment]
Looks good to me!  Great work!

**Comment by glasnt** [comment]
The merge from master and this branch aren't working properly. Hold off merging til I get 'er fixed.

**Comment by glasnt** [comment]
That should have fixed it.

File: pigeonhole.rb [hunk 5]
```diff
@@ -35,6 +36,12 @@
   redirect "/noise-candidates/#{today}/#{today}"
 end
 
+def search_precondition
+  return "" unless @search
```
Comment (nerdyness) [inline pigeonhole.rb]: Try to prefer ' over " wherever possible.

**Comment by nerdyness** [comment]
Happy for you to merge this :+1: :shipit:

**Pull request merged**